src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,278 @@
+using Sentry.Extensibility;
+using Sentry.Internal.Extensions;
+using Sentry.Protocol.Metrics;
+
+namespace Sentry;
+
+internal class MetricAggregator : IMetricAggregator, IDisposable
+{
+    internal enum MetricType : byte { Counter, Gauge, Distribution, Set }
+
+    private readonly SentryOptions _options;
+    private readonly Action<IEnumerable<Metric>> _captureMetrics;
+    private readonly TimeSpan _flushInterval = TimeSpan.FromSeconds(5);
+
+    private readonly CancellationTokenSource _shutdownSource;
+    private volatile bool _disposed;
+
+    // The key for this dictionary is the Timestamp for the bucket, rounded down to the nearest RollupInSeconds... so it
+    // aggregates all of the metrics data for a particular time period. The Value is a dictionary for the metrics,
+    // each of which has a key that uniquely identifies it within the time period
+    internal ConcurrentDictionary<long, ConcurrentDictionary<string, Metric>> Buckets => _buckets.Value;
+    private readonly Lazy<ConcurrentDictionary<long, ConcurrentDictionary<string, Metric>>> _buckets
+        = new(() => new ConcurrentDictionary<long, ConcurrentDictionary<string, Metric>>());
+
+    private Task LoopTask { get; }
+
+    public MetricAggregator(SentryOptions options, Action<IEnumerable<Metric>>? captureMetrics = null, CancellationTokenSource? shutdownSource = null, bool disableLoopTask = false)
+    {
+        _options = options;
+        _captureMetrics = captureMetrics ?? (_ => { });
+        _shutdownSource = shutdownSource ?? new CancellationTokenSource();
+
+        if (disableLoopTask)
+        {
+            // We can stop the loop from running during testing
+            _options.LogDebug("LoopTask disabled.");
+            LoopTask = Task.CompletedTask;
+        }
+        else
+        {
+            options.LogDebug("Starting MetricsAggregator.");
+            LoopTask = Task.Run(RunLoopAsync);
+        }
+    }
+
+    internal static string GetMetricBucketKey(MetricType type, string metricKey, MeasurementUnit unit, IDictionary<string, string>? tags)
+    {
+        var typePrefix = type switch
+        {
+            MetricType.Counter => "c",
+            MetricType.Gauge => "g",
+            MetricType.Distribution => "d",
+            MetricType.Set => "s",
+            _ => throw new ArgumentOutOfRangeException(nameof(type), type, null)
+        };
+        var serializedTags = tags?.ToUtf8Json() ?? string.Empty;
```

Comment:
Yeah, this is just replicating how the buckets work in Python... Within each time bucket, a "metric bucket" is a set of metrics for a unique combination of:
1. Metric type
2. Key provided by the user
3. Measurement unit type
4. Tags

The python code also serializes the tags to turn these into part of the composite key.

Potentially we could use a Hash of the JSON string... that would save space but increase computation. 

Alternatively, we could concatenate all the values/keys, using some kind of delimeter (which would have to be escaped if it was present in any of the values/keys). That might be slightly cheaper both computationally and in terms of storage.

Reply:
> Potentially we could use a Hash of the JSON string... that would save space but increase computation.

I was hoping we could avoid JSON encoding altogether. A cheaper way to stringify it might be enough for as long there's no chance of collision (`obj.GetHashCode()` is not a good choice for example)

> Alternatively, we could concatenate all the values/keys, using some kind of delimeter (which would have to be escaped if it was present in any of the values/keys). That might be slightly cheaper both computationally and in terms of storage.

Yeah something like this could work. We can look at this later, it's not a blocker here at all, but we need to benchmark this stuff to avoid regressing on performance after we had to situations recently of folks pointing out high CPU/Mem usage with Sentry.